doc/design/drafts/dynamic_size_types.md

```diff
@@ -0,0 +1,562 @@
+does. Function developers for instance do not want to restrict their functions
+to shared memory allocated types which would result in some implementation
+overhead. One example could be that someone analyses an image with a function
+like
+`void analyseImage(cxx::vector<int, SharedMemoryAllocator> & a)`. This function
+can now be only used with a vector which is using the `SharedMemoryAllocator`.
+One could define the allocator type as a template but this would increase the
+compile time and can make it nearly impossible to write proprietary code since
+templates require the implementation in the header. Furthermore, the function
+developer has to be aware of the implementation detail that there is something
+like a `SharedMemoryAllocator`.
+
+## Memory Layout
+
+In the following section we would like to send the structure 
+
+```cpp
+struct MyData {
+  int         value;
+  vector<int> image;
+  string      text;
+};
+```
+
+and observe how this is stored currently in the shared memory of iceoryx and
+compare it to the suggested solution.
+
+An iceoryx container memory structure looks mostly like this:
+```
+vector/string {
+  mgmt; // contains internal variables to handle the logic and later the
+        // suggested allocator
+
+  data; // contains the user data which is stored in the container
+}
+```
+
+### Static Layout
+
+When we require a sample with our publisher and write `MyData` we acquire a
+chunk where all the contents are stored inside one chunk. Hereby is the
+Memory Manager the class which handles the distribution of shared memory chunks.
+To support different sizes the Memory Manager uses bucket allocators with
+preconfigured bucket sizes called Memory Pool.
+
+```
++--------------------- Memory Manager -------------------+
+|  Memory Pool with chunk size = 1000                    |
+|  +----------------+----------------+----------------+  |
+|  |  value         |                |                |  |
+|  |  image::mgmt   |                |                |  |
+|  |  image::data   |                |                |  |
+|  |  text::mgmt    |                |                |  |
+|  |  text::data    |                |                |  |
+|  +----------------+----------------+----------------+  |
+|                                                        |
+|  Memory Pool with chunk size = 100                     |
+|  +---+---+---+---+---+---+---+---+---+---+---+---+---+ |
+|  |   |   |   |   |   |   |   |   |   |   |   |   |   | |
+|  |   |   |   |   |   |   |   |   |   |   |   |   |   | |
+|  |   |   |   |   |   |   |   |   |   |   |   |   |   | |
+|  +---+---+---+---+---+---+---+---+---+---+---+---+---+ |
++--------------------------------------------------------+
+```
+
+### Layout With Dynamic Sized Types
+
+With dynamic sized types we move the data part outside of the container and
+add an allocator to the container which allows the container to acquire
+memory for their data part. When we assume that `MyData::image` and
```

Comment:
Please describe this in more detail. The data part will now be stored in separate chunks and the management part will be inside one chunk, now including (a pointer to) the allocator, right?

Reply:
@FerdinandSpitzschnueffler I added a more detailed description. Can you please check it out.